tensorflow/c/eager/gradient_checker.cc

```diff
@@ -0,0 +1,200 @@
+
+// Fills data with values [start,end) with given step size.
+void Range(int data[], int start, int end, int step = 1) {
+  for (int i = start; i < end; i += step) {
+    data[i] = i;
+  }
+}
+
+// Returns AbstractTensorHandlePtr containing [0, ..., n-1].
+AbstractTensorHandlePtr GetRangeTensorHandleUtil(AbstractContext* ctx, int n) {
+  int vals[n];
+  int64_t vals_shape[] = {n};
+  Range(vals, 0, n);
+  AbstractTensorHandlePtr r = GetTensorHandleUtilInt(ctx, vals, vals_shape, 1);
+  return r;
+}
+
+// Fills out_dims with the dimensions of the given tensor.
+void GetDims(const TF_Tensor* t, int64_t* out_dims) {
+  int num_dims = TF_NumDims(t);
+  for (int i = 0; i < num_dims; i++) {
+    out_dims[i] = TF_Dim(t, i);
+  }
+}
+
+// Runs Model and reduce_sums the output.
+Status RunModelAndSum(AbstractContext* ctx, Model forward,
+                      absl::Span<AbstractTensorHandle*> inputs,
+                      absl::Span<AbstractTensorHandle*> outputs,
+                      bool use_function) {
+  GradientRegistry registry;
+  std::vector<AbstractTensorHandle*> model_outputs(1);
+
+  // Run the model.
+  TF_RETURN_IF_ERROR(RunModel(forward, ctx, inputs,
+                              absl::MakeSpan(model_outputs), use_function,
+                              registry));
+  AbstractTensorHandle* f_toSum = model_outputs[0];
+
+  TF_Tensor* model_out_tensor;
+  TF_RETURN_IF_ERROR(GetValue(f_toSum, &model_out_tensor));
+  int num_dims_out = TF_NumDims(model_out_tensor);
+
+  // Will sum all dimensions, so get a Tensor containing [0,...,num_dims_out-1].
+  AbstractTensorHandlePtr sum_dims =
+      GetRangeTensorHandleUtil(ctx, num_dims_out);
+
+  // Reduce sum the output on all dimensions.
+  std::vector<AbstractTensorHandle*> sum_inputs(2);
+  sum_inputs[0] = f_toSum;
+  sum_inputs[1] = sum_dims.get();
+
+  TF_RETURN_IF_ERROR(ops::Sum(ctx, absl::MakeSpan(sum_inputs),
+                              absl::MakeSpan(model_outputs), "sum_output"));
+  outputs[0] = model_outputs[0];
+  return Status::OK();
+}
+
+// Runs model as is if output is a scalar,
+// else sums the output tensor before returning.
+Status RunAndMaybeSum(AbstractContext* ctx, Model forward,
+                      absl::Span<AbstractTensorHandle*> inputs,
+                      absl::Span<AbstractTensorHandle*> outputs,
+                      bool use_function, bool is_scalar_out) {
+  Status s;
+  if (is_scalar_out) {
+    GradientRegistry registry;
+    s = RunModel(forward, ctx, inputs, outputs, use_function, registry);
+  } else {
+    s = RunModelAndSum(ctx, forward, inputs, outputs, use_function);
+  }
+  return s;
+}
+// ========================= End Helper Functions==============================
+
+Status CalcNumericalGrad(AbstractContext* ctx, Model forward,
+                         std::vector<AbstractTensorHandle*> inputs,
+                         float* dtheta_approx, int input_index,
+                         bool use_function, bool is_scalar_out) {
```

Comment:
For simplicity, should we assume that `Model` always outputs a single scalar tensor? That way you can get rid of the `is_scalar_out` arg.

Reply:
Sure, I can put a disclaimer in the description that the output must be a scalar for the gradient checker to work.

Reply:
@saxenasaurabh  Actually thinking about this a little more should we leave this as is except make `is_scalar_out` an optional parameter that defaults to true? That way a user doesn't have to specify it explicitly if their model already outputs a scalar, but has the option to still use this API if not. Otherwise we're putting the onus on the client to add a step to their existing models instead of being able to run it as is. What do you think?

Reply:
Another option is to have a function signature like 
```Status CalcNumericalGrad(AbstractContext* ctx, Model forward,
                         std::vector<AbstractTensorHandle*> inputs,
                         float* dtheta_approx, int input_index,
                         bool use_function)
```
that just calls the existing CalcNumericalGrad with `is_scalar_out = true`

Reply:
Just updated this to get rid of is_scalar_out; now we sum the output if the number of dimensions of the model's output is > 0.